### PR TITLE
Add alternative paths for CUDA installation.

### PR DESCRIPTION
**Pull request by pwnall**
This detects negativo17's CUDA packages for Fedora.

I tried following the feedback in #15614, so I added alternative paths and avoided adding questions to the `./configure` script.

### CODE REVIEW

**Comment by pwnall** [comment]
So far, this PR is one step short of fully working. I haven't figured out how to set the `NVVMIR_LIBRARY_DIR` environment variable to the directory returned by `_find_nvvm_libdevice_dir`. After setting that environment variable manually in `.bazelrc`, I ended up with a fully working CUDA build on Fedora 27 with negativo17's drivers.

Even if there's no easy way to set the environment variable, I figure there's value in getting the rest of the work reviewed and merged, so others have less customizing to do. I look forward to your feedback!

File: third_party/gpus/cuda_configure.bzl [hunk 1]
```diff
@@ -526,6 +526,27 @@ def _find_cuda_lib(lib, repository_ctx, cpu_value, basedir, version="",
...
+  the other CUDA libraries but rather in a special extras/CUPTI directory.
+
+  Args:
+    repository_ctx: The repository context.
+    cuda_config: The CUDA config as returned by _get_cuda_config
+
+  Returns:
+    The path of the directory containing the cupti header.
+  """
+  cuda_toolkit_path = cuda_config.cuda_toolkit_path
```
Comment (gunan) [inline third_party/gpus/cuda_configure.bzl]: Let's make it easier to add new alternate paths like this here:

```
cupti_alternate_paths = [ ....... ] # We can even define this list as a constant at the top of this file.

for path in cupti_alternate_paths:
  if repository_ctx.path(cuda_toolkit_path + path).exists:
    return ....

auto_configure_fail....
```

Same comment for all the modified methods.

**Review by gunan — CHANGES_REQUESTED** [review]
Same comment throughout the file.

**Comment by martrs** [comment]
negativo17 rpm package cuda already sets NVVMIR_LIBRARY_DIR so configure.py could read that and if it exists, call write_to_bazelrc(), or give user a choice by calling set_build_var(environ_cp, 'NVVMIR_LIBRARY_DIR',..). The same with cuda_configure.bzl.

Anyway, I can confirm the patch builds without problem on Fedora 27.

**Comment by tsoernes** [comment]
I was able to get past the initial path issues on Fedora 27 with Negativo drivers by 
```
export NVVMIR_LIBRARY_DIR=\usr\share\cuda
bazel build --config=opt --config=cuda --action_env="NVVMIR_LIBRARY_DIR=${NVVMIR_LIBRARY_DIR}" //tensorflow/tools/pip_package:build_pip_package
```
However the build still fails later on:
[tens-build.txt](https://github.com/tensorflow/tensorflow/files/1713722/tens-build.txt)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by pwnall** [comment]
Sorry, the past few weeks have been rough. I'll try to find some time to address the comments this week.

File: third_party/gpus/cuda_configure.bzl (hunk 1 above)
Comment (pwnall) [reply third_party/gpus/cuda_configure.bzl]: Done.

Thank you for the suggestion, the file looks a lot saner now!

File: third_party/gpus/cuda_configure.bzl [hunk 2]
```diff
@@ -646,15 +699,31 @@ def _find_cudnn_header_dir(repository_ctx, cudnn_install_basedir):
...
     The path of the directory containing the cudnn header.
   """
-  if repository_ctx.path(cudnn_install_basedir + "/cudnn.h").exists:
-    return cudnn_install_basedir
-  if repository_ctx.path(cudnn_install_basedir + "/include/cudnn.h").exists:
-    return cudnn_install_basedir + "/include"
+  for relative_path in CUDA_INCLUDE_PATHS:
+    if repository_ctx.path("%s/%scudnn.h" % (cudnn_install_basedir, relative_path)).exists:
+        return ("%s/%s" % (cudnn_install_basedir, relative_path))[:-1]
   if repository_ctx.path("/usr/include/cudnn.h").exists:
```
Comment (pwnall) [inline third_party/gpus/cuda_configure.bzl]: This is a bit of a wrinkle. The alternatives I could think of are:

1) Leave as-is -- not worth adding complexity for a single special case
2) Add CUDA_INCLUDE_ABSOLUTE_PATHS
3) Support leading "/" in paths and add logic for handling them. At that point the per-function logic gets complex enough that I'd want to extract helpers.

WDYT?

**Comment by pwnall** [comment]
@gunan Thank you very much for your feedback! I hope I've addressed it. PTAL?

In case it makes your life easier, I've also created CL 187808905 with the changes here.

**Comment by gunan** [comment]
@pwnall has sent this PR internally, closing the external one.
Thanks for the contribution!

**Comment by pwnall** [comment]
@gunan Thank you very much for the quick responses!

Everyone else interested in this: The internal patch has been committed, and should show up soon.

**Pull request closed without merging**